### PR TITLE
Redshift lint create external table statements

### PR DESCRIPTION
**Pull request by tinder-albertyue**
### Brief summary of the change made

Makes progress on #2145. 

### Are there any other side effects of this change that we should be aware of?
I was not sure how redshift_keywords worked. Could use some advice on that

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by tinder-albertyue** [comment]
Additional tests incoming. Might get to it later today or tomorrow

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 1]
```diff
@@ -283,6 +283,81 @@ class CreateTableAsStatementSegment(BaseSegment):
...
+                # Columns and comment syntax:
+                Delimited(
+                    Sequence(
+                        Ref("ColumnReferenceSegment"),
+                        Ref("DatatypeSegment"),
+                    ),
+                ),
+            )
+        ),
+        OneOf(Ref("PartitionedBySegment"), optional=True),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_redshift.py:311]:
```suggestion
        Ref("PartitionedBySegment", optional=True)
```
you can mark the ref itself as optional

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 2]
```diff
@@ -283,6 +283,81 @@ class CreateTableAsStatementSegment(BaseSegment):
...
+        "STORED",
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_redshift.py:312]: We're missing this bit from the docs above here:
```
[ { ROW FORMAT DELIMITED row_format |
  ROW FORMAT SERDE 'serde_name' 
  [ WITH SERDEPROPERTIES ( 'property_name' = 'property_value' [, ...] ) ] } ]
```
Would you be interested in adding it in? If not you should add a TODO note to the end of the docstring to let future contributors know the grammar is incomplete.

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 3]
```diff
@@ -283,6 +283,81 @@ class CreateTableAsStatementSegment(BaseSegment):
...
+        "AS",
+        OneOf(
+            "PARQUET",
+            "RCFILE",
+            "SEQUENCEFILE",
+            "TEXTFILE",
+            "ORC",
+            "AVRO",
+            Ref("QuotedLiteralSegment"),
+        ),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_redshift.py:353]:
```suggestion
        OneOf(
            "PARQUET",
            "TEXTFILE",
        ),
```
From the docs:
`The CREATE EXTERNAL TABLE AS command only supports two file formats, TEXTFILE and PARQUET.`

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 4]
```diff
@@ -323,13 +398,16 @@ class StatementSegment(BaseSegment):
...
             Ref("ColumnEncodingSegment"),
-            Ref("CreateUserSegment"),
+            Ref("CreateExternalTableAsStatementSegment"),
+            Ref("CreateExternalTableStatementSegment"),
             Ref("CreateGroupSegment"),
-            Ref("AlterUserSegment"),
-            Ref("AlterGroupSegment"),
+            Ref("CreateUserSegment"),
+            Ref("PartitionedBySegment"),
+            Ref("TableAttributeSegment"),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_redshift.py:410]: Again you shouldn't be reordering things that are unrelated to your PR, it just makes the diffs unclear...

File: test/fixtures/dialects/redshift/redshift_create_external_table_as.sql [hunk 5]
```diff
@@ -0,0 +1,51 @@
+CREATE EXTERNAL TABLE
+external_schema.table_name
+STORED AS PARQUET
+LOCATION 's3://bucket/folder/'
+AS
+SELECT col1, col2 FROM external_schema.source_table
+;
```
Comment (jpy-git) [inline test/fixtures/dialects/redshift/redshift_create_external_table_as.sql:7]: in the grammar you mention that this can be optionally bracketed, so in one of the examples could you make this:
```
AS (
    SELECT col1, col2 FROM external_schema.source_table
)
;
```

**Review by jpy-git — COMMENTED** [review]
@tinder-albertyue couple things to look at, shaping up well so far though 👍

File: src/sqlfluff/dialects/dialect_redshift.py (hunk 3 above)
Comment (tinder-albertyue) [reply src/sqlfluff/dialects/dialect_redshift.py:353]: good catch, thanks!

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 6]
```diff
@@ -283,6 +283,81 @@ class CreateTableAsStatementSegment(BaseSegment):
...
+        "STORED",
+        "AS",
+        OneOf(
+            "PARQUET",
+            "RCFILE",
+            "SEQUENCEFILE",
+            "TEXTFILE",
+            "ORC",
+            "AVRO",
+            Ref("QuotedLiteralSegment"),
```
[the remark this answers is not in the document]
Comment (tinder-albertyue) [reply src/sqlfluff/dialects/dialect_redshift.py:321]: I see, yea I misinterpreted the docs

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 7]
```diff
@@ -283,6 +283,81 @@ class CreateTableAsStatementSegment(BaseSegment):
...
+
+    type = "create_external_table_statement"
+
+    match_grammar = Sequence(
+        "CREATE",
+        "EXTERNAL",
+        "TABLE",
+        Ref("TableReferenceSegment"),
+        OneOf(Ref("PartitionedBySegment"), optional=True),
+        "STORED",
```
[the remark this answers is not in the document]
Comment (tinder-albertyue) [reply src/sqlfluff/dialects/dialect_redshift.py:347]: added in https://github.com/sqlfluff/sqlfluff/pull/2229/commits/3be449b963b45e8ef78b80dc923463b403187c6d

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 8]
```diff
@@ -283,6 +283,81 @@ class CreateTableAsStatementSegment(BaseSegment):
...
+            "RCFILE",
+            "SEQUENCEFILE",
+            "TEXTFILE",
+            "ORC",
+            "AVRO",
+            Ref("QuotedLiteralSegment"),
+        ),
+        "LOCATION",
+        Ref("QuotedLiteralSegment"),
+    )
```
[the remark this answers is not in the document]
Comment (tinder-albertyue) [reply src/sqlfluff/dialects/dialect_redshift.py:328]: The scope of TABLE PROPERTIES is fairly large. I'll add a TODO

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 9]
```diff
@@ -283,6 +283,81 @@ class CreateTableAsStatementSegment(BaseSegment):
...
+            "RCFILE",
+            "SEQUENCEFILE",
+            "TEXTFILE",
+            "ORC",
+            "AVRO",
+            Ref("QuotedLiteralSegment"),
+        ),
+        "LOCATION",
+        Ref("QuotedLiteralSegment"),
+        "AS",
```
[the remark this answers is not in the document]
Comment (tinder-albertyue) [reply src/sqlfluff/dialects/dialect_redshift.py:355]: added TODO

File: src/sqlfluff/dialects/dialect_redshift.py (hunk 2 above)
Comment (tinder-albertyue) [reply src/sqlfluff/dialects/dialect_redshift.py:310]: The scope is a bit daunting with the many possible properties and values. I will add a TODO and potentially address in a future PR

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -16,13 +16,14 @@ repos:
...
-          types-pkg_resources,
-          types-chardet,
           types-appdirs,
+          types-backports,
+          types-chardet,
           types-colorama,
+          types-pkg_resources,
           types-pyyaml,
           types-regex,
+          types-toml,
```
[the remark this answers is not in the document]
Comment (tinder-albertyue) [reply .pre-commit-config.yaml]: makes sense, reverted for now and will make a separate PR for this change

File: src/sqlfluff/dialects/dialect_redshift.py (hunk 4 above)
Comment (tinder-albertyue) [reply src/sqlfluff/dialects/dialect_redshift.py]: Makes sense, reverted the ordering

**Comment by tinder-albertyue** [comment]
@jpy-git thanks for the detailed review! I think I have addressed all the points and the pr is ready for another look

File: .pre-commit-config.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (jpy-git) [reply .pre-commit-config.yaml]: awesome thanks 👍

File: src/sqlfluff/dialects/dialect_redshift.py (hunk 2 above)
Comment (jpy-git) [reply src/sqlfluff/dialects/dialect_redshift.py:312]: Yeah there's no pressure to add everything at once and we can build up the grammar incrementally 😄

**Review by jpy-git — APPROVED** [review]
@tinder-albertyue fantastic work, thanks for adding this! LGTM 🚀 ✨
